### ISSUE TITLE
Internalise HTTP extensions (remove dependency on HTTPlease)

### ISSUE THREAD

**Issue by tintoy** [tintoy/dotnet-kube-client#166]
As discussed in [#165](https://github.com/tintoy/dotnet-kube-client/pull/165#issuecomment-2503667764), copy a minimum subset of functionality (e.g. URI templates, and consistent request/response handling) into dotnet-kube-client (e.g. in the `KubeClient` assembly under the `KubeClient.Http` namespace) to maintain source-level compatiblity (although not binary compatibility).

Where possible, use existing .NET extensions

Initial targeting should be for `net6.0`, `net7.0`, `net8.0`, `net9.0`.

## Some initial notes

Possible (partial) replacement (of implementation) for HTTPlease URI-template functionality:

* https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.extensions.urihelper?view=aspnetcore-6.0
* https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.extensions.querybuilder?view=aspnetcore-6.0

(note: these look impractical to consume, since the only way to get access to them would require a `<FrameworkReference Include="Microsoft.AspNetCore.App" />` from the KubeClient project that uses them)

Alternative implementation (and one of the most popular) of URI templates:

https://github.com/tavis-software/Tavis.UriTemplates

**Comment by tintoy** [comment]
Have split the HTTP client functionality out into a separate package `KubeClient.Http` that is an implicit dependency of the main `KubeClient` package.

Later, we can look at trimming this implementation (once the dust has settled from any breaks in compatibility) but for now it's pretty close to ready for release. I'm hoping to publish the new version of KubeClient in the next week or 2.

@raman-m - would you be interested in trying out the new version before I publish anything (to see if it's likely to cause any problems for you)? If so, I can make pre-release packages available on the project's MyGet feed.

**Comment by raman-m** [comment]
>would you be interested in trying out the new version before I publish anything (to see if it's likely to cause any problems for you)? If so, I can make pre-release packages available on the project's MyGet feed.

Yes, I would be willing to collaborate on testing in January, as I am currently focused on releasing the Ocelot .NET 9 version.

> Otherwise, I don't mind trying to pull down the Ocelot repository and try the upgrade myself (if you don't mind answering any questions that come up while I try to get it going).

Sounds reasonable. Why not wait for Ocelot v24.0, which will open a window for us to test both libraries? 
Anyway, you can use the actual code in the [develop branch](https://github.com/ThreeMammals/Ocelot/tree/develop) to clone, compile, and test. Currently, it is unclear to me what tests should be provided from the Ocelot side?

**Comment by tintoy** [comment]
> Yes, I would be willing to collaborate on testing in January, as I am currently focused on releasing the Ocelot .NET 9 version.

Great - thanks!

Basically, as long as it turns out that I’ve maintained source-code compatibility (excluding minor namespace changes) and behavioural compatibility then I think we’re good to go. Most of my concerns around the changes are what the upgrade experience will be like for existing consumers (which I can use when adding a migration guide to the documentation).  I can wait til v24.0, but will also try out the develop branch in the meanwhile. I’ll see what existing tests there are in there that relate to the area where KubeClient is being used; since we now have a [TestKit](https://github.com/tintoy/dotnet-kube-client/blob/84b055c885a2afe00781559ea400c0bd8cdfce6d/test/KubeClient.Extensions.DataProtection.Tests/KeyPersistenceTests.cs#L57) for KubeClient there may be some options there too.

**Comment by tintoy** [comment]
BTW, I’m planning on switching from JSON.NET to System.Text.Json eventually; is this likely to cause problems for you?

**Comment by raman-m** [comment]
> is this likely to cause problems for you?

It depends. But seems we have independent referencing; thus, no problems should occur.
But FYI there is open PR https://github.com/ThreeMammals/Ocelot/pull/2125, so we plan to support both `Newtonsoft.Json` and `System.Text.Json` packages until Microsoft no longer supports the [Microsoft.AspNetCore.Mvc.NewtonsoftJson](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.NewtonsoftJson) package.
